pints/_mcmc/_slice_rank_shrinking.py

```diff
@@ -0,0 +1,265 @@
+    and by definying a Markov chain.
+
+    Sampling follows:
+
+    1. Calculate the pdf (``f(x_0)``) of the current sample (``x_0``).
+    2. Draw a real value (``y``) uniformly from (0, f(x0)), defining a
+       horizontal “slice”: S = {x: y < f(x)}. Note that ``x_0`` is
+       always within S.
+    3. Draw the first crumb (``c_1``) from a Gaussian distribution with
+       mean ``x_0`` and precision matrix ``W_1``.
+    4. Draw a new point (``x_1``) from a Gaussian distribution with mean
+       ``c_1`` and precision matrix ``W_2``.
+
+    New crumbs are drawn until a new proposal is accepted. In particular,
+    after sampling ``k`` crumbs from Gaussian distributions with mean ``x0``
+    and precision matrices (``W_1``, ..., ``W_k``), the distribution for the
+    ``k`` th proposal sample is:
+
+    1. ``x_k \sim Normal(\bar{c}_k, \Lambda^{-1}_k)``
+
+    where:
+
+    2. ``\Lambda_k = W_1 + ... + W_k``
+    3. ``\bar{c}_k = \Lambda^{-1}_k * (W_1 * c_1 + ... + W_k * c_k)``
+
+    This method aims to conveniently modify the ``k + 1`` th proposal
+    distribution to increase the likelihood of sampling an acceptable point.
+    It does so by calculating the gradient (``g(f(x))``) of the unnormalised
+    posterior (``f(x)``) at the last rejected point (``x_k``).
+    It then sets the conditional variance of the ``k + 1`` proposal
+    distribution in the direction of the gradient ``g(f(x_k))`` to 0.
+    This is reasonable in that the gradient at a proposal probably points in
+    a direction where the variance is small, so it is more efficient to move
+    in a different direction.
+
+    To avoid floating-point underflow, we implement the suggestion advanced
+    in [1] pp.712. We use the log pdf of the un-normalised posterior
+    (``log f(x)``) instead of ``f(x)``. In doing so, we use an
+    auxiliary variable ``z = log(y) - \epsilon``, where
+    ``\epsilon \sim \text{exp}(1)`` and define the slice as
+    S = {x : z < log f(x)}.
+
+    [1] Thompson, M. and Neal, R.M., 2010. Covariance-adaptive slice sampling.
+    arXiv preprint arXiv:1003.3201.
+
+    *Extends:* :class:`SingleChainMCMC`
+    """
+
+    def __init__(self, x0, sigma0=None):
+        super(SliceRankShrinkingMCMC, self).__init__(x0, sigma0)
+
+        # Set initial state
+        self._x0 = np.asarray(x0, dtype=float)
+        self._running = False
+        self._ready_for_tell = False
+        self._current = None
+        self._current_log_y = None
+        self._proposed = None
+
+        # Standard deviation of initial crumb
+        self._sigma_c = 1
+
+        # Matrix of orthonormal columns of directions in which the conditional
+        # variance is zero
+        self._J = np.zeros((self._n_parameters, 0), float)
+
+        # Number of crumbs
+        self._k = 0
+
+        # Mean of unprojected proposal distribution
+        self._c_bar = 0
+
+    # Function returning the component of vector v orthogonal to the
+    # columns of J
+    def P(self, J, v):
```

Comment:
Since this is a function unlikely to be of use to users, should it be _P? @MichaelClerx

Reply:
Since its not doing anything with the object, it might be a ``_p`` (no capitals if possible) outside of the class (but in the same file)